### PR TITLE
default to using 'physical cores / 2' number of threads

### PR DESCRIPTION
**Pull request by elindsey**
I know there's some history here, so I'm not trying to reopen _too_ big of a can of worms, but... 🙂 

Immediately asking users to make a decision that they're not equipped to make isn't a great initial experience, especially when we have a recommended default. I understand that the optimal worker count will vary by workload and will likely require tuning, but it doesn't seem worthwhile to introduce friction at the very start of the process - especially when the most likely result of that friction is simply passing the default 'cores / 2' config and carrying on. A better alternative would be to log a visible warning if the default is used, pointing to the relevant tuning docs.

I also think there's a big difference between requiring no startup parms and >=1 startup params. The extremely minimal config materialize takes to standup is a fairly unique benefit in this space and does lead to a great initial UX; I'd hate to give that up unless truly necessary.

About the default itself - physical cores / 2 seems fine, but if we wanted a more conservative config to cast a wider net for potential 'Docker on a laptop' scenarios then defaulting to 1 core if 4 cores or less and cores / 2 above that also seems good.

(not going to invest time into the test failures yet)

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4407)

### CODE REVIEW

File: src/materialized/src/bin/materialized.rs
Comment (umanwizard) [inline src/materialized/src/bin/materialized.rs:203]: Presumably we need to round this up to 1 on single-core machines

**Review by umanwizard — COMMENTED** [review]
I have no strong desire to bikeshed whether this is a good idea, so I'll let others handle that! But if we _do_ go with this approach, what do you think about at least printing out a hint telling people this is configurable?

**Comment by rjnn** [comment]
@awang you should make a final decision here

**Comment by awang** [comment]
Linking the longer previous discussion:  https://github.com/MaterializeInc/materialize/pull/1726#issuecomment-600114856

**Comment by cuongdo** [comment]
I believe we should go with a reasonable default. `number of physical cores / 2` seems like a fine starting point. 1 core is probably not, because it violates a common user expectation that server software makes use of multiple cores when they are available. `number of physical cores` is probably too high, because we have Tokio and Kafka client thread pools, which aren't included in the worker count.

Here's a quick survey of parallelism for RDBMS software:
* PostgreSQL uses a process per client and set of background processes
* MySQL can use a thread per client connection or a thread pool
* Oracle has multiple sets of processes, including one for clients (I'm oversimplifying)

**Comment by awang** [comment]
Dumb question:  Why isn't this instead `(number of physical cores-1) / 2` (with a floor of 1, obviously)?  (I'm referring to https://github.com/MaterializeInc/materialize/pull/1726#issuecomment-579791430)

**Comment by awang** [comment]
Dumb question:  Why isn't this instead `(number of physical cores-1) / 2` (with a floor of 1, obviously)?  (I'm trying to address the `num_cpus - 1` comment from https://github.com/MaterializeInc/materialize/pull/1726#issuecomment-579791430) Ie:

| Number of Physical Cores  | Default Worker Setting |
| ------------- | ------------- |
| 1  | 1  |
| 2  | 1  |
| 3 | 2  |
| 4 | 3  |

**Comment by awang** [comment]
Other than my question above, I'm good with this change.

**Comment by elindsey** [comment]
`(number of physical cores-1) / 2` would be

| Number of physical cores | Default worker count|
|---|---|
|1|1|
|2|1|
|3|1|
|4|1|
|5|2|
|6|2|

So it skews more conservative. That's good if we want to overindex on making the out of the box laptop experience solid, but doesn't address the gist of the linked comment which is about a higher default for VMs. I think `num/2` is a good compromise that we can extend later on if necessary (by doing more detection of the two competing environments)

**Comment by elindsey** [comment]
Forgot to take this out of draft. 🙂 

Rebased and fixed the one broken UT. Per last comment stayed with cores/2, but let me know if you'd prefer (cores-1)/2 and I can make that change before the merge.

File: src/materialized/tests/cli.rs
Comment (quodlibetor) [inline src/materialized/tests/cli.rs:35]: these two seem like they should still be errors, as materialize won't work with zero workers.

Comment (elindsey) [reply src/materialized/tests/cli.rs:35]: Oof thanks for the catch, I was overzealous

**Comment by quodlibetor** [comment]
The code LGTM and I agree that requiring `-w` is kind of annoying.

**Comment by quodlibetor** [comment]
Actually, something worth checking: have you observed if `num_cpus` correctly notices docker limits on docker for mac? Like if docker only has access to 2 cores on a 16-core mac, will materialized start with 1 or 8 workers with this patch?

**Comment by elindsey** [comment]
> Actually, something worth checking: have you observed if num_cpus correctly notices docker limits on docker for mac? Like if docker only has access to 2 cores on a 16-core mac, will materialized start with 1 or 8 workers with this patch?

It correctly notices and will start 1 worker in that case. The only quirk with docker for mac is the issue where VMs lose context on physical versus hyper threaded cores, so if you devote all 8 logical cores of a quad core hyper threaded CPU to docker then it'll start up 4 workers. But we can't save users from themselves if they provision like that.

**Review by benesch — APPROVED** [review]
Sweet! Can you also update the docs in cli/_index.md to mark that this is no longer required and add a release note? (Feel free to do in a follow-up PR for faster CI.)

**Pull request merged**
